### ISSUE TITLE
Test failures on Fedora

### ISSUE THREAD

**Issue by jamesjer** [arminbiere/kissat#2]
Since cvc4 1.8 adds support for kissat, I am looking into creating a Fedora package of kissat.  The tests fail on 3 of our supported architectures.

On the two 32-bit architectures (arm and i686), the 2 GB allocation on line 20 of testallocate.c fails:
```
  p = kissat_calloc (solver, 1 << 29, 4);
```
The code doesn't check that p is nonnull, so the test segfaults when p is dereferenced on line 23.  Simply changing all of the 29s to 28s leads to a successful run.  Are you willing to make that change?  (If not, I can simply patch those files during the Fedora build.)

On the big endian architecture (s390x), testendianess.c fails, as it is written to do.  Is the issue that you don't have access to big endian hardware and therefore are unable to test?  If so, and if I can figure out the code changes necessary to make the other tests pass, are you willing to look at a patch?

I have an unrelated question, which I can open a separate issue for if need be: are you willing to look at a patch that adds the ability to build a shared library?

**Comment by arminbiere** [comment]
OK, probably assuming that everybody these days has a machine where this succeeds was overly optimistic.
I changed those three constants from 29 to 28.

Regarding shared libraries.  I remember that the same issue occurred for CaDiCaL but then Mathias solved the shared linking part outside of the CaDiCaL code.  The only trick necessary was to allow to pass '-fPIC ...' to the linker (which should work with Kissat in the same way as with CaDiCaL).  Does it not?

**Comment by arminbiere** [comment]
I am trying to test via cross-compilation and QEMU, but this might take some time.

**Comment by jamesjer** [comment]
> OK, probably assuming that everybody these days has a machine where this succeeds was overly optimistic.
> I changed those three constants from 29 to 28.

It's very likely that nobody will run kissat, or cvc4, on 32-bit hardware.  We like to keep building for those platforms anyway, because they do have their uses, and because their tighter resource constraints often turn up bugs sooner than they would manifest on 64-bit platforms.  That's not the case here, of course.

> Regarding shared libraries. I remember that the same issue occurred for CaDiCaL but then Mathias solved the shared linking part outside of the CaDiCaL code. The only trick necessary was to allow to pass '-fPIC ...' to the compiler via a configure option (which should work with Kissat in the same way as with CaDiCaL by using './configure -fPIC && make'). Does it not?

Yes, I just wondered if you would accept a patch to add a makefile target for building a shared library.  That way, nobody has to bother you about it in the future, should other Linux distributions decide to include kissat.  If you don't want the maintenance burden, I understand.

> The endianess is tricky. When I decided to use little endian layout for unions of bit-fields I feared that this will at point in the future become an issue (as you say, this is the purpose of that particular test). I wanted to avoid bit-macros as in Lingeling, to make the code more readable, while at the same time being forced that bits of different structs in those unions are at the same place. This is also very deeply connected to some assumptions of memory layout (the main point of Kissat) in the design of the code. I fear I will need to look at this myself. But I do not have access to a big-endian machine.

QEMU is slow.  I don't mean to distract you from more pressing concerns.  I have access to s390x hardware, if you would like me to look at the issue.  I may even be able to get you an account on that machine, if you are interested.

**Comment by arminbiere** [comment]
I tried already yesterday cross compilation (SPARC,PPC) and QEMU testing without problem, and as expected the endianess test and other related tests fail (for traversal of watch lists), but then I did not bother to really start working on this (also GDB will probably not easy to use even though I did not look into that yet).  I will stick to that for the moment and come back to you if this ends up being too slow - I ported once Lingeling and other solvers to IBM main frame for one QBF competition and back then they did not even have GCC support (and debugging was pretty tough: learned that in 'struct { int bit : 2 }' the signedness of 'bit' is undefined in C and worse signed for GCC and unsigned for IBM's compiler - so you explicitly need to specify sigendness for bit fields).

I want to leave the shared library building (and maintenance of that part) to those who need it (same with CMAKE).  Reasoning is that if you are using Kissat (or CaDiCaL) for a project in which you need dynamic libraries (or CMAKE) then you will need a more complicated build set-up anyhow.

**Comment by arminbiere** [comment]
I think I managed to port to big-endian.  There is only the watch union which seems problematic with respect to little and big endian layout of bit-fields.  I rely on the GCC and Clang supported macro __BYTE_ORDER__ to detect endianess (first added code to the configure script but removed that now).  In essence (as already envisioned in the 'testendianess.c' test case) one just has to mirror the bit-fields, also because those important single, and double bits (and explicit padding) are less than a byte. For testing this I use the following:

CC=s390x-linux-gnu-gcc ./configure -static && make tissat
qemu-s390x ../build/tissat
...
Running 685 jobs in parallel using up to 12 processes.

All 685 test jobs succeeded in 7.92 seconds.

So factor 5 slower than natively which is think is OK. I also checked with PowerPC, and Sparc64.  The latter has a strange directory issue, but the actual solving works.

Can you test this on a real big-endian machine, the IBM?

**Comment by arminbiere** [comment]
This works for me and thus I close it with the sc2021 release.
